### PR TITLE
Unify compression handling

### LINKED ISSUE #542: Unify compression handling
- [x] all backends should throw unsupported encoding exception when given encoding is not supported (looking at you `HttpClientBackend`)
- [ ] there should be a way to enrich given backend with custom encoding
- [ ] because of the previous point we shouldn't relay on automatic encoding handling in particular backends (e.g. okHttpClient and gzip)

This relates to #531

### PR DESCRIPTION
**Pull request by ghostbuster91**
This fixes #542

### CODE REVIEW

File: httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientBackend.scala [hunk 1]
```diff
@@ -106,17 +110,22 @@ abstract class HttpClientBackend[F[_], S](client: HttpClient, closeClient: Boole
...
+    val byteBody = if (method != Method.HEAD) {
+      customEncodingHandler.orElse(PartialFunction.fromFunction(standardEncoding.tupled))(res.body(), encoding)
     } else {
       res.body()
     }
     val body = responseHandler(byteBody).handle(responseAs, responseMonad, responseMetadata)
     responseMonad.map(body)(Response(_, code, "", headers, Nil))
   }
 
+  private def standardEncoding: (InputStream, Option[String]) => InputStream = {
```
Comment (adamw) [inline httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientBackend.scala:122]: shouldn't this be `(InputStream, String) => InputStream`? That is, the EncodingHandler would be only to specify behavior when there *is* any encoding?

No-encoding would be the default case.

Comment (ghostbuster91) [reply httpclient-backend/src/main/scala/sttp/client/httpclient/HttpClientBackend.scala]: Right, fixed.

**Comment by ghostbuster91** [comment]
I couldn't find a way to intercept automatic decompression in AHC and finagle

File: core/src/test/scala/sttp/client/testing/HttpTest.scala [hunk 2]
```diff
@@ -283,14 +285,20 @@ trait HttpTest[F[_]]
       val req = basicRequest.head(uri"$endpoint/compress")
       req.send().toFuture().map { resp => resp.code shouldBe StatusCode.Ok }
     }
+//TODO investigate as it breaks the rest of the tests
```
Comment (ghostbuster91) [inline core/src/test/scala/sttp/client/testing/HttpTest.scala:288]: Uncommenting this test causes tests to hang

Comment (adamw) [reply core/src/test/scala/sttp/client/testing/HttpTest.scala:288]: no idea :D which test hangs? maybe it's something in js/native?

Comment (ghostbuster91) [reply core/src/test/scala/sttp/client/testing/HttpTest.scala:288]: From the tests' output it seems like the one with big timeout

**Comment by ghostbuster91** [comment]
Also I am not sure if AkkaHttp's encodingHandler should have type `(HttpResponse, HttpEncoding) => HttpResponse` or just `HttpResponse => HttpResponse` as `HttpEncoding` can be extracted from `HttpResponse`.

File: core/src/test/scala/sttp/client/testing/HttpTest.scala (hunk 2 above)
Comment (ghostbuster91) [reply core/src/test/scala/sttp/client/testing/HttpTest.scala:288]: it hangs jvm tests. I didn't run native and js locally.

**Comment by eugeniyk** [comment]
@adamw wonder why it was reverted?

**Comment by ghostbuster91** [comment]
Because of some failing tests, eventually it was merged into master, but as another PR AFAIR

**Pull request merged**